### ISSUE TITLE
[Federated] Services namespaces

### ISSUE THREAD

**Issue by terion-name** [apollographql/federation#350]
Currently I have a system built with microservices and with a self-written gateway that uses schema-stitching and a lot of magic.
This system was written several years ago and now needs a refresh and Apollo Federation has most of features that I've implemented by my own but better designed. Now I want to migrate to Federation but it seems to a problem.

My usecase maybe specific but I think it is not uniquely rare.

_**In short**: services should have namespaces._

Namespaces are needed to prevent type collisions. The idea behind is that services are autonomous, are written by numerious developers, including external ones. And services can (and do) have alike types (e.g. `News` has `Article` and `Blog` has `Article`). 

Controlling types uniqueness "administratively" is not an option due to big amount of services and people involved. Also considering auto-generated CRUD types (e.g. from `Prisma`) this is becoming even more of a problem. To 100% eliminate type collisions and bring a unified convention to end-users gateway should namespace services. 

In practice it looks like this:

Services:
```graphql
# news service
Query {
   articles: [Article!]!
}
Article {
  id
  body
}
```
```graphql
# blog service
Query {
   articles: [Article!]!
}
Article {
  id
  content
}
```

At the gateway schemas are transforming before stitiching and result is like this:
```graphql
Query {
   News {
      articles: [NewsArticle!]!
   }
   Blog {
     articles: [BlogArticle!]!
   }
}
NewsArticle {
  id
  body
}
BlogArticle {
  id
  content
}
```

This approach proven to be convenient with tens or hundreds of services with hundreds or thousands of types and with autogeneration on services level. 

Looking at Apollo Federation I don't see a possibility to do this. So the questions:

1. Maybe there is a possibility to transform service schemas that I've overlooked?
2. Maybe such a feature should be considered like an option? Something like:
```js
new ApolloGateway({
  serviceList: [
    { name: 'news', url: 'http://localhost:4001', namespace: 'News' },
    { name: 'blog', url: 'http://localhost:4001', namespace: 'Blog' }
  ],
});
```

**Comment by kevin-lindsay-1** [comment]
Doesn't `extend type Name @key` give you enough information to detect that a particular type exists outside of the schema's namespace?

For example:

```graphql
# "Loan" service

type Subject implements Node { # The subject of a loan (typically an individual or a business)
  id: ID!
  ...
}

extend type Subject implements Node @key("id") { # The "global" variant of subject, the Auth version of a subject
  id: ID!
 ... @external
}
```

When running `buildFederatedSchema` couldn't you detect these via the fact that the type has `extend` and `@key` on it, and prep the typedefs for dynamic namespacing via codegen in `buildFederatedSchema`?

As an example, you might see the type names get converted to something machine-readable like `Subject` and `__EXTERNAL__Subject`, respectively.

Then, in the gateway, you add the service, with an option for the `namespace: "loan"` and those type definitions can then be renamed to something more appropriate, such as `LoanSubject` and `Subject`, respectively.

---

While this idea might work, it might not necessarily be the easiest to read, and it might be difficult to understand at a glance from which service this type originates.

Maybe it would be useful for the sake of preventing global namespace collision and readability to specify something like:

```grahql
[extend] type Subject implements Node @external("auth") @key("id") { # similar to declaring external fields, you can declare the type as external and the name of the service it will use in federation
  id: ID!
  ... @external
}
```

This might make manually declaring a service's "namespace" irrelevant, as the gateway should theoretically have enough information to detect and resolve namespace collision this way, with the caveat of it theoretically being necessary to declare an "authoritative" type if a collision occurs.

For example, let's say `Subject` (auth) exists, and then `Subject` (loan) comes along. It would be theoretically be useful in that instance to be able to state that the auth version was part of the original API, so as to not introducing a breaking change. A directive like "`@primary`" might be useful in that instance, which would tell the gateway that said type definition is the one that should be able to live in the non-namespaced scope.

Whether or not the `@primary` type should also create a namespaced variant is a different question, as I imagine some users wouldn't tag any type definitions as `@primary` until a global namespace conflict arises, at which point you can resolve the conflict by declaring one as `Subject @primary(deprecation: Date)`, and then `Subject` would appear in the federated schema as `@deprecated("Transition to <Namespace>Subject by <Date>")`.

**Comment by richardscarrott** [comment]
Yeah, it would be interesting to hear more about how people are solving this. We built our API gateway using schema stitching ~3 years ago and just blanket namespaced all types + root fields, e.g.

```
    const rawSchema = await getRemoteSchema(usersLink);
    return transformSchema(rawSchema, [
      new RenameTypes(name => `Users${ucFirst(name)}`),
      new RenameRootFields((operation, name) => `users${ucFirst(name)}`),
    ]);
```

It's worked fine, but makes our API a bit grim to read, e.g.

```
query Stuff {
   usersUser(id: "usr_123") {
      __typename // UsersUser
      id
     locale {
         __typename // UsersLocale
         region
         language
         currency
     }
   }
   ordersOrder(id: "odr_123") {
      __typename // OrdersOrder
      id
      locale {
         __typename // OrdersLocale
         region
         language
         currency
     }
      user {
         __typename // UsersUser
         id
      }
   }
}
```

**Comment by korinne** [comment]
Hi all, I'm the Product Manager for the Apollo Federation Core team 👋 We're currently researching this request, and we'd love to hear any more feedback/requirements for namespaces in Federation. 

There's a lot of great detail in this thread, but we wanted to open the door for more candid feedback -- where we can dive into your specific use-cases and concerns. With that, if you'd like to discuss this issue further, please feel free to either schedule a call directly on my [Calendly link](https://calendly.com/korinne-alpers/chat-with-the-apollo-federation-team), or email me at korinne@apollographql.com. Thanks in advance!
